src/server/tweetFinder.js

```diff
@@ -1,63 +1,85 @@
-import { flattenImmutableObject } from '../shared/utilities';
+import { chainPromises, flattenImmutableObject } from '../shared/utilities';
+import { TwitAccess, searchAndSaveFromTwitter } from './twitterSearch';
 
 /**
- * Grab all tweets from the database, and show them with their authors.
- * @deprecated
- * @param response
+ * Searches our database for Tweets and returns them.
+ * If our search is not good enough, access Twitter directly to retrieve more tweets.
+ * @param req A HTTP Request object
+ * @param res A HTTP Response object
+ * @returns {Promise.<T>|*}
  */
-export const exampleDatabaseCall = (request, response) => {
-  db.query('SELECT FROM tweet WHERE content LUCENE :query LIMIT 20',
-    {
-      'params': {
-        'query': request.params.query + '~',
+export const searchQuery = (req, res) => {
+  const query = req.body[0].query;
+  return searchAndCollateResults(query)
+    .then(
+      (data) => {
+        res.writeHead(200, { 'Content-Type': 'application/json' });
+        res.end(JSON.stringify(data));
       },
-    })
-    .then((tweetRecords) => {
-      const result = [];
-
-      const promises = tweetRecords.map((tweetRecord) => {
-        const rid = '' + tweetRecord['@rid'];
-
-        return db.query(
-          `SELECT FROM (TRAVERSE in(TWEETED) FROM (SELECT FROM ${rid})) WHERE @class = "Tweeter"`
-          )
-          .then((tweeterRecords) => {
-            const tweeterRecord = tweeterRecords[0];
-            const toReturn = {
-              'tweet':
-                flattenImmutableObject(
-                  TweetBuilder()
-                    .id(tweetRecord.id)
-                  .content(tweetRecord.content)
-                  .date(tweetRecord.date.toISOString())
-                  .likes(tweetRecord.likes)
-                  .retweets(tweeterRecord.retweets)
-                  .build()),
-              'tweeter':
-                flattenImmutableObject(
-                  TweeterBuilder()
-                    .id(tweeterRecord.id)
-                  .name(tweeterRecord.name)
-                  .handle(tweeterRecord.handle)
-                  .build()),
-            };
-
-            result.push(toReturn);
-          });
-      });
-
-      Promise.all(promises)
-        .then(() => {
-          response.end(
-            JSON.stringify(
-              result));
+
+      (rejection) => {
+        res.writeHead(500, { 'Content-Type': 'application/json' });
+        res.end('An unexpected internal error occurred.');
+        console.warn(`Unable to search for query '${query}'`, rejection);
+      }
+    );
+};
+
+const searchAndCollateResults = (query) => {
+  return searchDatabase(query)
+    .then((data) => {
+      return {
+        'data': {
+          'count': data.length,
+          'records': getTweetsAsResults(data),
+        },
+      };
+    });
+};
+
+const searchDatabase = (query, alreadyAttemptedRefresh = false) => {

```

Comment:
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: ['searchDatabase' was used before it was defined.](https://www.codacy.com/app/winwardo/solid-octo-disco/file/2180581610/issues/source?bid=3148782&fileBranchId=3164498#l42)